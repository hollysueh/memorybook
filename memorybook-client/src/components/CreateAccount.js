import React, { Component } from 'react';
import { Form } from 'react-bootstrap';
import cookie from 'js-cookie';
import '../App.css';

//Create Account Page
class CreateAccount extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formUsername: '',
      formEmail: '',
      formPassword: ''
    };
    this.getUser = this.getUser.bind(this);
    this.getPwd = this.getPwd.bind(this);
    this.handleCreateAccount = this.handleCreateAccount.bind(this);
    this.handleLogin = this.handleLogin.bind(this);
  }

  //Change states depending on what text was entered into form
  getUser(e) {
    this.setState({
      formUsername: e
    });
  }
  getEmail(e) {
    this.setState({
      formEmail: e
    });
  }
  getPwd(e) {
    this.setState({
      formPassword: e
    });
  }

  //Send form values to server to create new user account
  handleCreateAccount(e) {
    e.preventDefault(); //stop page reloading before fetch has completed
    fetch('/newUser', { //send POST request to server
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      }, // send server 'create account' form values
      body: JSON.stringify({
        user: this.state.formUsername,
        email: this.state.formEmail,
        pwd: this.state.formPassword
      }),
    })
    .then(res => res.json())
    .then(
      (result) => {
        console.log(result);
        if (result === "User already exists") { //if user already has an account...
          alert('Oops! You already have an account with us. Please try logging in.'); //alert user that there is an existing account
          window.location.href = "./Login"; //redirect page to 'Login' page
        } 
        else if (result === "Some error occured") { //if there was an error creating the account...
          alert('Oops! It looks like there was a problem. Please try again.'); //alert user that there was an error
        } 
        else { //if user account successfully created...
          alert('Account created!'); //alert user that account was successfully created
          this.handleLogin(); //log in new user
        }
    });
  }

  //Log In user
  handleLogin() {
    //Once new user account has been created, send "login" request to server so user is logged in
    fetch('/login', { //send POST request to server
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      }, // send server 'create account' form values
      body: JSON.stringify({
        user: this.state.formUsername,
        email: this.state.formEmail,
        pwd: this.state.formPassword
      }),
    })
    .then(res => res.json())
    .then(
      (result) => {
        if (result === "err: Incorrect login") { //if login details are incorrect...
          alert('Oops! An error occured. Please try again.'); //alert user that login was unsuccessful
        } else { //if login details are correct...
          //Set cookie as users authorization token 
          cookie.set('token', result.token, {expires: 1});
          window.location.href = `./PhotoAlbum/${result.userID}`; //redirect page to My Memory Book' page, send userID as params
        }
    });
  }

  //Render Create Account form
  render() {
    return (
      <div className="App">
        <h3 className="pageHeading">Create a new Account</h3>
        <link
          rel="stylesheet"
          href="https://maxcdn.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css"
          integrity="sha384-GJzZqFGwb1QTTN6wy59ffF1BuGJpLSa9DkKMp0DgiMDm4iYMj70gZWKYbI706tWS"
          crossOrigin="anonymous"/>
        <Form className="accountForm">
          <Form.Group>
            <Form.Label className="formLabel">Username:</Form.Label>
            <Form.Control type="text" placeholder="username" onChange={(e) => this.getUser(e.target.value)}/>
            <br></br>
            <Form.Label className="formLabel">Email:</Form.Label>
            <Form.Control type="text" placeholder="email" onChange={(e) => this.getEmail(e.target.value)}/>
            <br></br>
            <Form.Label className="formLabel">Password:</Form.Label>
            <Form.Control type="text" placeholder="password" onChange={(e) => this.getPwd(e.target.value)}/>
            <br></br>
          </Form.Group>
          <br></br>
          <button className="accountButton" variant="primary" type="login" onClick={this.handleCreateAccount} >
          Create Account
          </button>
        </Form>
        <br></br>
        <p>Already have an account? <a href="/Login">Login here.</a></p>
      </div>
    );
  }
}

export default CreateAccount;