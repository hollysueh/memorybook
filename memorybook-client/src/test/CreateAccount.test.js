import React from 'react';
import CreateAccount from '../components/CreateAccount';
import renderer from 'react-test-renderer';

//Test CreateAccount.js renders correctly
test('renders correctly', () => {
  const tree = renderer
  .create(<CreateAccount />)
  .toJSON();
  expect(tree).toMatchSnapshot();
});

//Test CreateAccount.js renders correctly when states populated
test('renders states correctly when populated', () => {
  const formUsername = 'Jane Doe';
  const formEmail = 'janedoe@gmail.com';
  const tree = renderer
  .create(<CreateAccount formUsername={formUsername} formEmail={formEmail} />)
  .toJSON();
  expect(tree).toMatchSnapshot();
});

//Test fetch() function used in CreateAccount.js is working correctly
test('create new user', () => {
  fetch('/newUser', { //send POST request to server
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    }, // send server 'account' values
    body: JSON.stringify({
      user: 'Jane Doe',
      email: 'janedoe@gmail.com',
      pwd: 'snapshotTest'
    }),
  })
  .then(res => {
    expect(res[0].artistName).toBeDefined();
  });
});
