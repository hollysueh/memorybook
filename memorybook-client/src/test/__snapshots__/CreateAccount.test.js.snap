// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <h3
    className="pageHeading"
  >
    Create a new Account
  </h3>
  <link
    crossOrigin="anonymous"
    href="https://maxcdn.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css"
    integrity="sha384-GJzZqFGwb1QTTN6wy59ffF1BuGJpLSa9DkKMp0DgiMDm4iYMj70gZWKYbI706tWS"
    rel="stylesheet"
  />
  <form
    className="accountForm"
  >
    <div
      className="form-group"
    >
      <label
        className="formLabel form-label"
      >
        Username:
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="username"
        type="text"
      />
      <br />
      <label
        className="formLabel form-label"
      >
        Email:
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="email"
        type="text"
      />
      <br />
      <label
        className="formLabel form-label"
      >
        Password:
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="password"
        type="password"
      />
      <br />
    </div>
    <br />
    <button
      className="accountButton"
      onClick={[Function]}
      type="login"
      variant="primary"
    >
      Create Account
    </button>
  </form>
  <br />
  <p>
    Already have an account? 
    <a
      href="/Login"
    >
      Login here.
    </a>
  </p>
</div>
`;

exports[`renders states correctly when populated 1`] = `
<div
  className="App"
>
  <h3
    className="pageHeading"
  >
    Create a new Account
  </h3>
  <link
    crossOrigin="anonymous"
    href="https://maxcdn.bootstrapcdn.com/bootstrap/4.2.1/css/bootstrap.min.css"
    integrity="sha384-GJzZqFGwb1QTTN6wy59ffF1BuGJpLSa9DkKMp0DgiMDm4iYMj70gZWKYbI706tWS"
    rel="stylesheet"
  />
  <form
    className="accountForm"
  >
    <div
      className="form-group"
    >
      <label
        className="formLabel form-label"
      >
        Username:
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="username"
        type="text"
      />
      <br />
      <label
        className="formLabel form-label"
      >
        Email:
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="email"
        type="text"
      />
      <br />
      <label
        className="formLabel form-label"
      >
        Password:
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="password"
        type="password"
      />
      <br />
    </div>
    <br />
    <button
      className="accountButton"
      onClick={[Function]}
      type="login"
      variant="primary"
    >
      Create Account
    </button>
  </form>
  <br />
  <p>
    Already have an account? 
    <a
      href="/Login"
    >
      Login here.
    </a>
  </p>
</div>
`;
